File: zig/ui/glfw/index_buffer.zig
const std = @import("std");
const gl = @cImport(@cInclude("GL/glew.h"));

pub const IndexBuffer = struct {
    rendererID: u32 = 0,
    count: u32 = 0,


    pub fn destroy(self: *IndexBuffer) void {
        gl.glDeleteBuffers().?(1, &self.rendererID);
    }

    pub fn bind(self: *IndexBuffer) void {
        gl.glBindBuffer().?(gl.GL_ELEMENT_ARRAY_BUFFER, self.rendererID);
    }

    pub fn unbind(self: *IndexBuffer) void {
        _ = &self;
        gl.glBindBuffer().?(gl.GL_ELEMENT_ARRAY_BUFFER, 0);
    }
};


pub fn indexBuffer(data: []u32) IndexBuffer {
    var ib: IndexBuffer = .{};
    comptime if (@sizeOf(u32) != @sizeOf(gl.GLuint)) {
        @panic("u32 not equalse GLuint");
    };

    gl.glGenBuffers().?(1, &ib.rendererID);
    gl.glBindBuffer().?(gl.GL_ELEMENT_ARRAY_BUFFER, ib.rendererID);
    gl.glBufferData().?(gl.GL_ELEMENT_ARRAY_BUFFER, @intCast(@sizeOf(u32) * data.len), data.ptr, gl.GL_STATIC_DRAW);
    ib.count = @intCast(data.len);

    return ib;
}

File: c++/ui/opengl/index_buffer.cpp
#include "index_buffer.hpp"

IndexBuffer::IndexBuffer(unsigned *data, unsigned size) {
    static_assert(sizeof(unsigned) == sizeof(GLuint));

    glGenBuffers(1, &this->rendererID);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, this->rendererID);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER, size, data, GL_STATIC_DRAW);
    this->count = size / sizeof(typeof(*data));
}

IndexBuffer::~IndexBuffer(){
    glDeleteBuffers(1, &this->rendererID);
}

void IndexBuffer::bind() {
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, this->rendererID);

}
void IndexBuffer::unbind() {
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
}
